### DISCUSSION TITLE
How to Dynamically Add Metadata to JWT Tokens in Supabase for Same User in Different Platforms

### DISCUSSION THREAD

**Discussion by Mohammed-Yasin-Mulla** [supabase/supabase#20732] [Questions]
I am working on a project with two platforms, a Mobile App and an Admin App, both using Supabase for user authentication. I want to control data access for users based on the platform they are signed in from, using Row Level Security (RLS) policies in Supabase.

My goal is to add metadata to the JWT token when a user is created, specifying the platform they are signing in from. I've attempted to achieve this by including the metadata in the options section during user sign-in, as shown below:
```typescript
  const { data: OTPData, error: OTPError } =
    await supabaseClient.auth.signInWithOtp({
      phone: `91${phoneNo}`,
      options: {
        shouldCreateUser: true,
        channel: "whatsapp",
        data: {
          app_metadata: { x: "admin" }, // Adding Meta Data here
        },
      },
    });
```
However, I've observed that this approach only adds metadata to new users in the authentication table, and the metadata is included in the JWT token regardless of whether the data option is provided or not.

I'm seeking guidance on how to dynamically add metadata to JWT tokens for users who have already been created, so that I can implement different RLS policies based on the platform they are signing in from. Any insights or suggestions on achieving this would be greatly appreciated.

**Comment by GaryAustin1** [comment]
Have you seen it documented that signInWithOtp handles the data option?   I notice in gotrue-js code it does pass it on, but it is not documented that I could find.  Normally signin does not take metadata option (like signInWithEmail) so it is very possible Otp does not update metadata if the user exists.  And since it is not documented you should not count on it for a new user with that method without confirmation from Supabase (maybe issue in gotrue repository) that it will always work.

If you are going to use the jwt method you might consider this library https://github.com/supabase-community/supabase-custom-claims and making a separate rpc call to set your info.  Do not the warning about the jwt not reflecting on the client without refreshing the session from the client.

Also all of this is not secure including your code above if run on a browser, but you might be OK if native app code.

**Reply by Mohammed-Yasin-Mulla** [reply]
Thank you @GaryAustin1  for your insightful response.

Indeed, the absence of the data field in the official documentation for `signInWithOtp` raised questions, but upon examining the methods for creating a new user, the inclusion of the data field is evident. Confirming its presence in the code and its functionality has been reassuring.
![Screenshot from 2024-01-26 08-56-54](https://github.com/supabase/supabase/assets/55092280/c70e7b36-a9e4-4f41-be1e-4b6d974042b1)
I appreciate your recommendation regarding the use of the supabase-custom-claims repository for managing custom claims. I will certainly explore its capabilities.

Regarding the security aspect, I am aware of the caution around client-side code, and I appreciate your reminder. While Supabase does highlight the safety of the anon key in the client, I am mindful of potential vulnerabilities and continuously strive to adhere to best practices for secure coding. If you have any additional insights or specific considerations regarding security, I would value your guidance.
![Screenshot from 2024-01-26 09-37-12](https://github.com/supabase/supabase/assets/55092280/b367ae48-a3d9-4fb5-b8bf-7b9d242385da)

Thank you once again for sharing your expertise.

**Reply by GaryAustin1** [reply]
Yes, I noticed the same code in the source.  Just always be cautious of using undocumented code.  I can't imagine why they would pull the code for new users as the signup call takes it.  Did you confirm though if the user exists the metadata does NOT get inserted into auth users?  That is the type of thing that is likely because the normal sign in call does not pass it either or insert/update it.

There are many ways to solve your issue, so I'm sure something will work for you.

**Reply by Mohammed-Yasin-Mulla** [reply]
@GaryAustin1  I took a quick look at the library, and it turns out it stores the meta-data in the auth.user table. You can find the details [here](https://github.com/supabase-community/supabase-custom-claims#where-are-these-custom-claims-stored). 

So, here's the thing – when a user logs in from some other platform, this meta-data is applied  for all users, no matter where they're logging in from. But, I'm looking for a way to figure out which platform they're coming from. So I can serve up the right stuff for each one.

If you've got any ideas on how to make this work, please do share And Yes, I will be happy if I am being proven wrong. Let me know what you think. Thanks a bunch for the help

**Reply by Mohammed-Yasin-Mulla** [reply]
@GaryAustin1 I am more than grateful if you could share your thougths on what I could do to achinve my goal.
